contracts/Router.sol

```diff
@@ -170,62 +231,71 @@ contract Router is ReentrancyGuard {
+        }
+
         // Increase our internal measure of treeSupply by the amount sent in plus the amount sent to LP rewards
-        treeSupply = treeSupply.add(amountIn.div(PRECISION.sub(rewardsCut)));
+        // minus the amount burned
+        treeSupply = treeSupply.add(amountIn.div(PRECISION.sub(rewardsCut))).sub(totalInBurnPool);
 
         if (!hasTransferredOldReserveBalance) {
             // move oldReserveBalance to charity by reversing the code that computes the charityCutAmount
-            // https://github.com/WhalerDAO/tree-contracts/blob/master/contracts/TREEReserve.sol#L173-L175
             amounts[0] = 0;
             amounts[1] = oldReserveBalance.div(charityCut).mul(PRECISION.sub(rewardsCut));
             hasTransferredOldReserveBalance = true;
         } else {
             // send some of the reserveToken to charity
-            uint256 charityCutAmount = reserveTokenReceived.mul(charityCut).div(
-                PRECISION.sub(rewardsCut)
-            );
+	        uint256 charityCutAmount = totalPledged.mul(charityCut).div(PRECISION.sub(rewardsCut));
             reserveToken.safeIncreaseAllowance(address(omniBridge), charityCutAmount);
             omniBridge.relayTokens(address(reserveToken), charity, charityCutAmount);
         }
 
-        emit Rebase(totalPledged, numPledgers);
+        emit Rebase(totalPledged, numPledgers, totalInBurnPool, numBurners);
 
-        // Reset tracking variables
+        // Reset pledging and burning pools
         totalPledged = 0;
         numPledgers = 0;
-    }
-
-
-    function burnTREE(uint256 amount) external nonReentrant {
-        // Burn TREE for msg.sender. This doesn't update TREE.totalSupply() like TREE.reserveBurn()
-        tree.transferFrom(msg.sender, amount, address(0));
-
-        // Give reserveToken to msg.sender based on % of Tree supply burned ^ 1.25
+        totalInBurnPool = 0;
+        numBurners = 0;
 
-        // totalReserveTokens * (amount ^ 1.25) / (treeSupply ^ 1.25) =
-        // totalReserveTokens * (amount / treeSupply) ^ 1.25
-        uint256 amountToReceive = reserveToken.balanceOf(address(this)).mul(
-            amount.mul(Babylonian.sqrt(Babylonian.sqrt(amount)))).div(
-            treeSupply.mul(Babylonian.sqrt(Babylonian.sqrt(treeSupply))));
-
-        reserveToken.safeTransfer(msg.sender, amountToReceive);
-
-        // Since we cant call TREE.reserveBurn(), we have to track treeSupply ourselves.
-        treeSupply = treeSupply.sub(amountToReceive);
-
-        emit BurnTREE(msg.sender, amount, amountToReceive);
+        // Increase targetPrice after a successful rebase
+	    //
+        // targetPrice = targetPrice.mul(targetPriceMultiplier).div(PRECISION);
     }
 
-    function withdrawToken(address _token, address _to, uint256 _amount, bool max) external payable {
-        require(msg.sender == gov, "UniswapRouter: not gov");
+
+    function withdrawToken(address _token, address _to, uint256 _amount, bool max) external onlyGov {
```

Comment:
Something I was thinking about was adding a catch to prevent withdrawal of reserve token. That way, to be able withdraw reserve token you'd also have to set reserve token to another token

Reply:
Could it be as simple as `require(_token != reserveToken.address, "Cannot withdraw reserve token")`?

Reply:
That makes sense to me.